res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        inValueScale(value) {
+            // Hack to get exact center of pots to return 0.5
+            if (value > (this.max / 2)) {
+                return (value - 1) / (this.max - 1);
+            } else {
+                return value / (this.max + 1);
+            }
+        }
+        // map input in the XML file, not inValueScale
+        input(channel, control, value, _status, _group) {
+            this.inSetParameter(this.inValueScale(value));
+        }
+        outValueScale(value) { return value * this.max; }
+        output(value, _group, _control) {
+            this.send(this.outValueScale(value));
+        }
+
+        // common functions
+        // In most cases, you should not overwrite these.
+        inGetParameter() {
+            return engine.getParameter(this.group, this.inKey);
+        }
+        inSetParameter(value) {
+            engine.setParameter(this.group, this.inKey, value);
+        }
+        inGetValue() {
+            return engine.getValue(this.group, this.inKey);
+        }
+        inSetValue(value) {
+            engine.setValue(this.group, this.inKey, value);
+        }
+        inToggle() {
+            this.inSetValue(!this.inGetValue());
+        }
+
+        outGetParameter() {
+            return engine.getParameter(this.group, this.outKey);
+        }
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
+            /**
+            Override this method with a custom one to connect multiple Mixxx COs for a single Component.
+            Add the connection objects to the this.connections array so they all get disconnected just
+            by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+            different color depending on the state of different Mixxx COs. Refer to
+            SamplerButton.connect() and SamplerButton.output() for an example.
+            **/
+            if (this.group !== undefined &&
+                this.outKey !== undefined &&
+                this.output !== undefined &&
+                typeof this.output === "function") {
+                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            }
+        }
+        disconnect() {
+            for (const conn of this.connections) {
+                conn.disconnect();
+            }
+        }
+        trigger() {
+
+            for (const conn of this.connections) {
+                conn.trigger();
+            }
+        }
+        send(value) {
+            if (this.midi === undefined || this.midi[0] === undefined || this.midi[1] === undefined) {
```

Comment:
this.midi isn't mentioned elsewhere, can we have a block of expected members at the top of the class, even if they are undefined?  That way programmers can know what values they can overwrite.

also, midi[0] and midi[1] are hard to understand without reading the implementation, can we make this something like midi.channel and midi.control?

Reply:
The purpose of the midi values are explained pretty clearly and in-depth in the wiki article. We could document them using jsdoc comments if you insist.

Reply:
Now that I have a working link to the wiki I'll go through that.  Usually I prefer to keep documentation with the code so the docs stay up to date, but that can be a change we make later so we can leave it for now.